File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.php
<?php

class AdamicAdar {
    private array $adjacencyList = [];

    public function __construct() {
        $this->adjacencyList = [];
    }

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacencyList[$u])) {
            $this->adjacencyList[$u] = [];
        }
        if (!isset($this->adjacencyList[$v])) {
            $this->adjacencyList[$v] = [];
        }
        $this->adjacencyList[$u][$v] = true;
        $this->adjacencyList[$v][$u] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function adamicAdarScore(int $u, int $v): float {
        if (!isset($this->adjacencyList[$u]) || !isset($this->adjacencyList[$v])) {
            return 0.0;
        }

        $score = 0.0;
        foreach (array_keys($this->adjacencyList[$u]) as $node) {
            if (isset($this->adjacencyList[$v][$node])) {
                $degree = count($this->adjacencyList[$node]);
                if ($degree > 1) {
                    $score += 1.0 / log($degree);
                }
            }
        }
        return $score;
    }

    public function scoreAllPairs(): array {
        $scores = [];
        $nodes = array_keys($this->adjacencyList);

        for ($i = 0; $i < count($nodes); $i++) {
            for ($j = $i + 1; $j < count($nodes); $j++) {
                $u = $nodes[$i];
                $v = $nodes[$j];
                $score = $this->adamicAdarScore($u, $v);
                if ($score > 0) {
                    $scores[] = [
                        'pair' => [$u, $v],
                        'score' => $score
                    ];
                }
            }
        }

        usort($scores, function($a, $b) {
            return $b['score'] <=> $a['score'];
        });

        return $scores;
    }

    public function topKPredictions(int $k = 10): array {
        $scores = $this->scoreAllPairs();
        return array_slice($scores, 0, $k);
    }
}

?>

File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class AdamicAdar
{
    private Dictionary<int, HashSet<int>> adjacencyList;

    public AdamicAdar()
    {
        adjacencyList = new Dictionary<int, HashSet<int>>();
    }

    public void AddEdge(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u))
            adjacencyList[u] = new HashSet<int>();
        if (!adjacencyList.ContainsKey(v))
            adjacencyList[v] = new HashSet<int>();
        
        adjacencyList[u].Add(v);
        adjacencyList[v].Add(u);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public double AdamicAdarScore(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u) || !adjacencyList.ContainsKey(v))
            return 0.0;

        double score = 0.0;
        foreach (int node in adjacencyList[u])
        {
            if (adjacencyList[v].Contains(node))
            {
                int degree = adjacencyList[node].Count;
                if (degree > 1)
                {
                    score += 1.0 / Math.Log(degree);
                }
            }
        }
        return score;
    }

    public List<((int, int), double)> ScoreAllPairs()
    {
        var scores = new List<((int, int), double)>();
        var nodes = adjacencyList.Keys.ToList();

        for (int i = 0; i < nodes.Count; i++)
        {
            for (int j = i + 1; j < nodes.Count; j++)
            {
                int u = nodes[i];
                int v = nodes[j];
                double score = AdamicAdarScore(u, v);
                if (score > 0)
                {
                    scores.Add(((u, v), score));
                }
            }
        }

        return scores.OrderByDescending(x => x.Item2).ToList();
    }

    public List<((int, int), double)> TopKPredictions(int k = 10)
    {
        return ScoreAllPairs().Take(k).ToList();
    }
}